### PR TITLE
Add accessible NcSelect and NcSelectTags components

### PR DESCRIPTION
**Pull request by Pytal**
- [x] Implement new component with https://github.com/sagalbot/vue-select for improved multiselect accessibility
- [x] Port NcMultiselectTags
- [x] Allow automatic wrapping to be disabled
- [x] Filter by `displayName` and `subtitle` on user option object when `userSelect` is enabled
- [x] Migrate NcMultiselect usage in this library to new API
    - [x] NcSettingsSelectGroup
    - [x] NcTimezonePicker
- ~Document breaking changes~
- [x] Add icon svg and title support to NcListItemIcon
- [x] Restore original NcMultiselect and NcMultiselectTags
- [x] Add NcSelect and NcSelectTags components
  - [x] Deprecate NcMultiselect and NcMultiselectTags

<details>
<summary>Related changes</summary>

- [ ] Fix styles https://github.com/nextcloud/server/pull/35165
- [x] Fix eslint errors https://github.com/nextcloud/nextcloud-vue/pull/3474

</details>

> https://github.com/sagalbot/vue-select follows many accessibility best practices as described in https://vue-select.org/guide/accessibility.html but will require upstream pull requests to further enhance accessibility

### CODE REVIEW

**Comment by Pytal** [comment]
Ready for code review with interactive examples mainly for NcMultiselect and NcMultiselectTags available https://deploy-preview-3435--nextcloud-vue-components.netlify.app/#/Components/NcMultiselect

File: src/components/NcTimezonePicker/NcTimezonePicker.vue [hunk 1]
```diff
@@ -44,13 +44,8 @@ export default {
 	<NcMultiselect :value="selectedTimezone"
 		:options="options"
 		:multiple="false"
-		:group-select="false"
 		:placeholder="placeholder"
-		group-values="regions"
-		group-label="continent"
```
Comment (artonge) [inline src/components/NcTimezonePicker/NcTimezonePicker.vue:50]: Does this mean that we lose grouping values ?

File: src/components/NcListItemIcon/NcListItemIcon.vue [hunk 2]
```diff
@@ -125,11 +130,14 @@ export default {
 		},
 
 		/**
-		 * Icon class to be displayed at the end of the component
+		 * Object containing the material design icon `component`
+		 * and the descriptive `title`
+		 *
+		 * @type {{ component: import('vue').Component, title: string }}
```
Comment (skjnldsv) [inline src/components/NcListItemIcon/NcListItemIcon.vue:136]: Nope, let's not pass components as references. This create issues.
See https://github.com/nextcloud/activity/pull/981#discussion_r1005817299

**Review by skjnldsv — CHANGES_REQUESTED** [review]

File: src/components/NcListItemIcon/NcListItemIcon.vue [hunk 3]
```diff
@@ -30,27 +30,27 @@ It might be used for list rendering or within the multiselect for example
 <NcListItemIcon title="User 1" subtitle="Hidden subtitle because size is too small" :avatar-size="24" />
 ```
 ```vue
-<NcListItemIcon title="User 1" :avatar-size="44" icon="icon-user" />
+<NcListItemIcon title="User 1" :avatar-size="44" :icon="UserIcon" />
```
[the remark this answers is not in the document]
Comment (Pytal) [reply src/components/NcListItemIcon/NcListItemIcon.vue]: Added a separate `iconSvg` prop in https://github.com/nextcloud/nextcloud-vue/pull/3435/commits/687b9f3221785602feb104150b2370bf46a27fac

[the remark this answers is not in the document]
Comment (Pytal) [reply src/components/NcListItemIcon/NcListItemIcon.vue]: Restored icon class support and added a separate `iconSvg` prop in https://github.com/nextcloud/nextcloud-vue/pull/3435/commits/687b9f3221785602feb104150b2370bf46a27fac

File: src/components/NcListItemIcon/NcListItemIcon.vue [hunk 4]
```diff
@@ -84,7 +84,12 @@ It might be used for list rendering or within the multiselect for example
 
 		<!-- @slot use this slot to add a custom icon or actions -->
 		<slot />
-		<span v-if="hasIcon && !hasSlot" class="icon option__icon" :class="icon" />
+
+		<component :is="icon.component"
+			v-if="icon && !hasSlot"
+			class="option__icon"
+			:size="20"
+			:title="icon.title" />
```
[the remark this answers is not in the document]
Comment (Pytal) [reply src/components/NcListItemIcon/NcListItemIcon.vue]: Restored icon class support

File: src/components/NcListItemIcon/NcListItemIcon.vue [hunk 5]
```diff
@@ -125,11 +130,14 @@ export default {
...
-		 * Icon class to be displayed at the end of the component
+		 * Object containing the material design icon `component`
+		 * and the descriptive `title`
+		 *
+		 * @type {{ component: import('vue').Component, title: string }}
 		 */
 		icon: {
-			type: String,
-			default: '',
+			type: Object,
```
[the remark this answers is not in the document]
Comment (Pytal) [reply src/components/NcListItemIcon/NcListItemIcon.vue]: Restored

File: src/components/NcMultiselect/NcEllipsisedOption.vue [hunk 6]
```diff
@@ -45,23 +45,15 @@ export default {
 	},
 
 	props: {
-		option: {
-			type: [String, Object],
-			required: true,
-			default: '',
-		},
```
[the remark this answers is not in the document]
Comment (Pytal) [reply src/components/NcMultiselect/NcEllipsisedOption.vue:52]: Yes stopped being used at some point

File: src/components/NcMultiselectTags/NcMultiselectTags.vue [hunk 7]
```diff
@@ -122,76 +131,162 @@ export default {
...
+			<NcEllipsisedOption :name="getOptionLabel(option)"
+				:search="search" />
 		</template>
-		<template #option="scope">
-			{{ tagLabel(scope.option) }}
+		<template #selected-option="selectedOption">
+			<NcEllipsisedOption :name="getOptionLabel(selectedOption)"
+				:search="search" />
+		</template>
+		<template v-for="(_, name) in $scopedSlots" #[name]="data">
```
[the remark this answers is not in the document]
Comment (Pytal) [reply src/components/NcMultiselectTags/NcMultiselectTags.vue:150]: The `data` here is populated by the `v-bind` in vue-select i.e. https://github.com/sagalbot/vue-select/blob/bbe0002f86d3beb18cb30b4bb418e66a7596b848/src/components/Select.vue#L116

File: src/components/NcTimezonePicker/NcTimezonePicker.vue (hunk 1 above)
Comment (Pytal) [reply src/components/NcTimezonePicker/NcTimezonePicker.vue:50]: Will revert some breaking changes discussed with @PVince81

**Comment by Pytal** [comment]
New NcSelect and NcSelectTags components ready for code review with interactive examples at https://deploy-preview-3435--nextcloud-vue-components.netlify.app/#/Components/NcSelect

File: src/components/NcSelect/NcSelect.vue [hunk 8]
```diff
@@ -0,0 +1,849 @@
...
+					return (option, label, search) => {
+						return (`${label} ${option.subtitle}` || '')
+							.toLocaleLowerCase()
+							.indexOf(search.toLocaleLowerCase()) > -1
+					}
+				}
+				return null
+			}
+			return this.filterBy
+		},
```
Comment (artonge) [inline src/components/NcSelect/NcSelect.vue:738]: Return early to ease readability ?

File: src/components/NcSelect/NcSelect.vue [hunk 9]
```diff
@@ -0,0 +1,849 @@
...
+
+		localLabel() {
+			if (this.label === null) {
+				if (this.userSelect) {
+					return 'displayName'
+				}
+				return 'label'
+			}
+			return this.label
+		},
```
Comment (artonge) [inline src/components/NcSelect/NcSelect.vue:748]: Return early to ease redability ?

File: src/components/NcSelectTags/api.js [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+	}
+	return result
+}
+
+const searchTags = async function() {
+	if (window.NextcloudVueDocs) {
+		return Promise.resolve(xmlToTagList(window.NextcloudVueDocs.tags))
+	}
+
+	const result = await axios({
```
Comment (artonge) [inline src/components/NcSelectTags/api.js:96]: You can use the `webdav` package, which will remove the need of the above functions I think:

Example in photos:
- https://github.com/nextcloud/photos/blob/master/src/services/DavClient.js
- https://github.com/nextcloud/photos/blob/master/src/services/Albums.js

**Review by artonge — CHANGES_REQUESTED** [review]

File: src/components/NcListItemIcon/NcIcon.vue [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+		title: {
+			type: String,
+			default: '',
+		},
+	},
+
+	data() {
+		return {
+			cleanSvg: '',
+			htmlString: '',
```
[the remark this answers is not in the document]
Comment (Pytal) [reply src/components/NcListItemIcon/NcIcon.vue]: Nope, as `import('@skjnldsv/sanitize-svg').sanitizeSVG` returns a promise and async computed properties were not implemented in Vue core, see https://github.com/vuejs/vue/issues/4083#issuecomment-259193445

File: src/components/NcListItemIcon/NcIcon.vue [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+				const titleElement = document.createElement('title')
+				titleElement.textContent = this.title
+				if (element.firstElementChild) {
+					element.firstElementChild.prepend(titleElement)
+				}
+			}
+
+			this.htmlString = element.outerHTML
+		},
+	},
```
[the remark this answers is not in the document]
Comment (Pytal) [reply src/components/NcListItemIcon/NcIcon.vue]: We don't expect broken SVGs but we need to parse the SVG to then add the accessible title element within the SVG element itself, see https://github.com/robcresswell/vue-material-design-icons/blob/5.1.2/build.ts#L26

File: src/components/NcSelectTags/api.js (hunk 10 above)
Comment (Pytal) [reply src/components/NcSelectTags/api.js:96]: Looks nice!

For the purposes of this PR though this was just copied from the existing https://github.com/nextcloud/nextcloud-vue/blob/v7.1.0-beta.1/src/components/NcMultiselectTags/api.js, so would be better to refactor this in a separate PR

**Review by artonge — APPROVED** [review]
Good for me.

The comment about having `NcIconSvgWrapper` public can be ignored for merging, as we can always do it later :)

**Comment by Pytal** [comment]
- [ ] Test component as sharing input in server

File: src/components/NcListItemIcon/NcIconSvgWrapper.vue [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+</template>
+
+<script>
+import { sanitizeSVG } from '@skjnldsv/sanitize-svg'
+
+import { t } from '../../l10n.js'
+import logger from '../../utils/logger.js'
+
+export default {
+	name: 'NcIconSvgWrapper',
```
[the remark this answers is not in the document]
Comment (Pytal) [reply src/components/NcListItemIcon/NcIconSvgWrapper.vue:38]: Any thoughts @skjnldsv?

[the remark this answers is not in the document]
Comment (skjnldsv) [reply src/components/NcListItemIcon/NcIconSvgWrapper.vue:38]: Yep, you can have <script> inside an svg. So We need to sanitise them :)

File: src/components/NcListItemIcon/NcIconSvgWrapper.vue [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+			if (this.title) {
+				const titleElement = document.createElement('title')
+				titleElement.textContent = this.title
+				if (element.firstElementChild) {
+					element.firstElementChild.prepend(titleElement)
+				}
+			}
+
+			this.htmlString = element.outerHTML
+		},
```
Comment (skjnldsv) [inline src/components/NcListItemIcon/NcIconSvgWrapper.vue:95]: Isn't it going too far?
We can just target the svg tag within via `.icon-vue svg`

**Review by skjnldsv — CHANGES_REQUESTED** [review]
see comments

File: src/components/NcListItemIcon/NcIconSvgWrapper.vue (hunk 13 above)
[the remark this answers is not in the document]
Comment (Pytal) [reply src/components/NcListItemIcon/NcIconSvgWrapper.vue:38]: 👍 already sanitized @skjnldsv, in https://github.com/nextcloud/nextcloud-vue/blob/enh/a11y-multiselect/src/components/NcListItemIcon/NcIconSvgWrapper.vue#L68

@artonge was wondering whether or not we should expose this component publicly?

File: src/components/NcListItemIcon/NcIconSvgWrapper.vue (hunk 14 above)
Comment (Pytal) [reply src/components/NcListItemIcon/NcIconSvgWrapper.vue:95]: The primary purpose of this snippet is to add the accessible title element similar to https://github.com/robcresswell/vue-material-design-icons/blob/5.1.2/build.ts and thought that we might as well add the class as good practice due to the increased specificity of classes over tags

Comment (skjnldsv) [reply src/components/NcListItemIcon/NcIconSvgWrapper.vue:95]: Not sure about performances though.
I think doing it from `update()` or `mounted()` and just appending the title or updating if already exists would be more efficient :)

But that can be a followup!

**Review by skjnldsv — APPROVED** [review]

**Pull request merged**